Without the fix, the test added in this PR fails with:

panic: interface conversion: parser.Datum is parser.DValArg, not parser.DInt

```
goroutine 375 [running]:
github.com/cockroachdb/cockroach/sql/parser.glob.func108(0x1, 0xece45b26d, 0x34929958, 0x21e6520, 0xece45b26d, 0x34929958, 0x21e6520, 0xc820506c00, 0xc8204bc280, 0xc8201681e0, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:150 +0xb5
github.com/cockroachdb/cockroach/sql/parser.(*BinaryExpr).Eval(0xc820620060, 0x1, 0xece45b26d, 0x34929958, 0x21e6520, 0xece45b26d, 0x34929958, 0x21e6520, 0xc820506c00, 0xc8204bc280, ...)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:816 +0x35f
github.com/cockroachdb/cockroach/sql.(*selectNode).renderRow(0xc820154480)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:481 +0x17f
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc820154480, 0x7f0bc5128320)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:125 +0x1ef
github.com/cockroachdb/cockroach/sql.(*subqueryVisitor).Visit(0xc8201af5b0, 0x7f0bc69b2ea8, 0xc8204bc720, 0xc8204bc901, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/subquery.go:122 +0x100d
github.com/cockroachdb/cockroach/sql/parser.WalkExpr(0x7f0bc40a8b98, 0xc8201af5b0, 0x7f0bc69b2ea8, 0xc8204bc720, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/parser/walk.go:210 +0x50
github.com/cockroachdb/cockroach/sql.(*planner).expandSubqueries(0xc8201ab000, 0x7f0bc69b2ea8, 0xc8204bc720, 0x1, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/subquery.go:26 +0xf1
github.com/cockroachdb/cockroach/sql.(*selectNode).addRender(0xc820154360, 0x7f0bc69b2ea8, 0xc8204bc720, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:424 +0xc3b
github.com/cockroachdb/cockroach/sql.(*selectNode).initTargets(0xc820154360, 0xc82062a140, 0x1, 0x1, 0xc820168210)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:307 +0xf2
github.com/cockroachdb/cockroach/sql.(*planner).initSelect(0xc8201ab000, 0xc820154360, 0xc8202e4140, 0x0, 0x0, 0xc8201bc780)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:168 +0x122
github.com/cockroachdb/cockroach/sql.(*planner).Select(0xc8201ab000, 0xc8202e4140, 0x0, 0x0, 0x7ffce386ec84)
	/go/src/github.com/cockroachdb/cockroach/sql/select.go:158 +0x7f
github.com/cockroachdb/cockroach/sql.(*planner).prepare(0xc8201ab000, 0x7f0bc40a8ac8, 0xc8202e4140, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/plan.go:167 +0x384
github.com/cockroachdb/cockroach/sql.(*Executor).Prepare(0xc8200c4000, 0xc82048800d, 0x4, 0xc82048904b, 0x14, 0xc8201681e0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/executor.go:240 +0x4d8
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleParse(0xc820676380, 0xc8206763b8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:322 +0x537
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc820676380, 0xc82062a000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:247 +0xf24
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serveConn(0xc8203ed4c0, 0x7f0bc40ecdb0, 0xc8204ee000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:219 +0xa15
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve.func1(0xc8203ed4c0, 0x7f0bc5128ad8, 0xc8201ec2b8)
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:126 +0x90
created by github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve
	/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:131 +0x20a
```

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4149#-:-K9i3yUx02x56uPRInS7:1756585305)*
<!-- Sent from Reviewable.io -->

--- a/acceptance/chaos_test.go
+++ b/acceptance/chaos_test.go
@@ -137,14 +137,11 @@ CREATE TABLE bank.accounts (
 					}
 					amount := rand.Intn(*maxTransfer)
 
-					// TODO(mjibson): We can't use query parameters with this query because
-					// it fails type checking on the CASE expression.
-					update := fmt.Sprintf(`
+					const update = `
 									UPDATE bank.accounts
-									  SET balance = CASE id WHEN %[1]d THEN balance-%[3]d WHEN %[2]d THEN balance+%[3]d END
-										  WHERE id IN (%[1]d, %[2]d) AND (SELECT balance >= %[3]d FROM bank.accounts WHERE id = %[1]d)
-											`, from, to, amount)
-					if _, err := clients[i].db.Exec(update); err != nil {
+									  SET balance = CASE id WHEN $1 THEN balance-$3 WHEN $2 THEN balance+$3 END
+										  WHERE id IN ($1, $2) AND (SELECT balance >= $3 FROM bank.accounts WHERE id = $1)`
+					if _, err := clients[i].db.Exec(update, from, to, amount); err != nil {
 						// Ignore some errors.
 						if testutils.IsError(err, "connection refused") {
 							return nil

--- a/sql/bank_test.go
+++ b/sql/bank_test.go
@@ -72,14 +72,11 @@ CREATE TABLE IF NOT EXISTS bench.bank (
 
 			amount := rand.Intn(*maxTransfer)
 
-			// TODO(mjibson): We can't use query parameters with this query because
-			// of the subquery.
-			update := fmt.Sprintf(`
+			const update = `
 UPDATE bench.bank
-  SET balance = CASE id WHEN %[1]d THEN balance-%[3]d WHEN %[2]d THEN balance+%[3]d END
-  WHERE id IN (%[1]d, %[2]d) AND (SELECT balance >= %[3]d FROM bench.bank WHERE id = %[1]d)
-`, from, to, amount)
-			if _, err := db.Exec(update); err != nil {
+  SET balance = CASE id WHEN $1 THEN balance-$3 WHEN $2 THEN balance+$3 END
+  WHERE id IN ($1, $2) AND (SELECT balance >= $3 FROM bench.bank WHERE id = $1)`
+			if _, err := db.Exec(update, from, to, amount); err != nil {
 				b.Fatal(err)
 			}
 		}

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -300,6 +300,9 @@ func TestPGPreparedQuery(t *testing.T) {
 		"SHOW TIME ZONE": {
 			base.Results("UTC"),
 		},
+		"SELECT (SELECT 1+$1)": {
+			base.Params(1).Results(2),
+		},
 		// TODO(mjibson): test date/time types
 	}
 
@@ -427,6 +430,12 @@ func TestPGPreparedExec(t *testing.T) {
 			"UPDATE d.t SET i = CASE WHEN $1 THEN i-$3 WHEN $2 THEN i+$3 END",
 			[]preparedExecTest{
 				base.Params(true, true, 3).RowsAffected(3),
+			},
+		},
+		{
+			"UPDATE d.t SET i = CASE i WHEN $1 THEN i-$3 WHEN $2 THEN i+$3 END",
+			[]preparedExecTest{
+				base.Params(1, 2, 3).RowsAffected(3),
 			},
 		},
 	}

--- a/sql/subquery.go
+++ b/sql/subquery.go
@@ -67,6 +67,10 @@ func (v *subqueryVisitor) Visit(expr parser.Expr, pre bool) (parser.Visitor, par
 		return nil, expr
 	}
 
+	if v.prepareOnly {
+		return nil, expr
+	}
+
 	if exists != nil {
 		// For EXISTS expressions, all we want to know is if there is at least one
 		// result.

--- a/sql/update.go
+++ b/sql/update.go
@@ -146,7 +146,7 @@ func (p *planner) Update(n *parser.Update) (planNode, *roachpb.Error) {
 			if _, ok := expr.(parser.DefaultVal); ok {
 				return nil
 			}
-			d, err := expr.TypeCheck(nil)
+			d, err := expr.TypeCheck(p.evalCtx.Args)
 			if err != nil {
 				return roachpb.NewError(err)
 			}

